VDB-1629 - Have CheckedHeadersRepository use the plugin schema specified

paytonrules opened the pull request:
As part of pulling Oasis into its own repository (and by extension removing it from MCD-transformers) we can't share the same "check_count" on the headers table anymore. Plugins provide their schema and must have a matching table.

This also removes the need for CLIENT_IPCPATH from the make test command, which was an accident.

paytonrules committed:
Inject the NewCheckedHeaders repo into LogExtractor

This makes no changes, but will let each plugin provide a way to check
their headers.

paytonrules committed:
Create a schema for header checking in test

The code doesn't use this yet, this just gets us a clean setup/teardown.

paytonrules committed:
Make the CHeckedHeadersRepository use a schema

The checked_headers table should not be in the public schema, it is
provided by the plugin with the expected name checked_headers.

paytonrules committed:
Explicitly call the query a joinQuery

Refactor - just a rename

paytonrules committed:
Pass unit tests without CLIENT_IPCPATH

The CLIENT_IPCPATH was being used by the
contract_watcher/repository/header_repository_test on accident. A helper
function it called happened to initialize a geth connection.

Remove that in favor of the simpler db config, and move the integration
test initialization and warning to integration_test/integration_test.

paytonrules committed:
Add a test for CLIENT_IPCPATH on integrationtest

paytonrules committed:
Extract SetupBC from SetupDBandBC

Only the SetupBC code was shared.

paytonrules committed:
Refactoring: Make the tests a little easier to read

paytonrules committed:
Fix bugs in two tests

Two tests should have been using a recheckCount of 3 - which is why they
didn't make sense when I tried to update them for the new query.
Updating them to a 3 means th checkCount is set to 2, and the block
number must be in the new adjusted window.

paytonrules committed:
Remove the check_count from headers

It does not belong there.

rmulhol commented on integration_test/contract_watcher_transformer_test.go:
```diff
@@ -29,13 +29,14 @@ import (
 var _ = Describe("contractWatcher transformer", func() {
 	var (
-		db               *postgres.DB
+		db               = test_config.NewTestDB(test_config.NewTestNode())
```
seems like we're overwriting this assignment in the `BeforeEach`?

paytonrules replied:
🤦

rmulhol commented on pkg/datastore/postgres/repositories/checked_headers_repository.go:
```diff
@@ -17,62 +17,77 @@
-	insertCheckedHeaderQuery = `UPDATE public.headers SET check_count = (SELECT check_count WHERE id = $1) + 1 WHERE id = $1`
+	insertCheckedHeaderQuery = `
+INSERT INTO %s.checked_headers (check_count, header_id)
+VALUES (1, $1)
```
I think we might want to retain the ability to mark the `check_count` as > 1, so that we can indicate rechecks

paytonrules replied:
We did - that's how the ON_CONFLICT WORKS.  See line 32 where there is the +1.

What has changed is that entries in the join table aren't present until the first check, whereas before we relied on the header row having a default check_count of 0. Now we try to insert a check row, and if it's present (ON CONFLICT) we increment it by one.

This should be the same behavior as before, as the only place I saw where we explicitly set a value was in tests. Otherwise it was the same  "increment by 1" logic.

rmulhol replied:
😅 I feel very roasted by you putting "works" in all caps 🔥

paytonrules replied:
😆 That was unintentional. I was just STILL TYPING THE SQL AND I GUESS KEPT GOING.

rmulhol commented on pkg/datastore/postgres/repositories/checked_headers_repository.go:
```diff
@@ -17,62 +17,77 @@
 
+	joinQuery := fmt.Sprintf(`SELECT h.id, h.block_number, h.hash
+						FROM public.headers h
+						LEFT JOIN %s.checked_headers ch
```
Just making sure I follow - since we're doing a left join from the headers table, our result will include any headers where no row currently exists on the checked_headers table. Is that right?

paytonrules replied:
Yes. We had good tests for this, fortunately, since I got it backwards the first time :).

rmulhol commented on pkg/datastore/postgres/repositories/checked_headers_repository.go:
```diff
@@ -17,62 +17,77 @@
+		noEndingBlockQuery := fmt.Sprintf(`%s
+				 WHERE ((ch.check_count IS NULL OR ch.check_count < 1) AND h.block_number >= $1)
+				 OR ((ch.check_count IS NULL OR ch.check_count < $2)
+			           AND h.block_number <= ((SELECT MAX(block_number) FROM public.headers) - ($3 * ch.check_count * (ch.check_count + 1) / 2)))`, joinQuery)
```
So seems like this may have been the case in the earlier implementation too, but does the second branch of this where clause disregard the starting block number?

paytonrules replied:
Looks like it does to me too. I hadn't noticed by I believe that's an error (albeit not a new one). Can try it out with a test.

paytonrules replied:
I cleaned up this query a bit while fixing this bug, but I fixed it and added some tests.

rmulhol commented on test_config/test_config.go:
```diff
@@ -47,16 +45,6 @@ func setTestConfig() {
-	}
-	if ipc == "" {
-		logrus.Fatal(errors.New("testing.toml IPC path or $CLIENT_IPCPATH env variable need to be set"))
-	}
```
😤

rmulhol reviewed (commented):
Looking good!

paytonrules committed:
Verify the configured schema actually exists

This should make SQL injection impossible, since we query for and use
that schema name, rather than using the one passed in by the user.

This does mean creating the new CheckedHeadersRepository can fail, so I
added an error, and I did some refactoring on the
checked_headers_repository_test because I learned about Expect / To(Succeed())

paytonrules committed:
Provide a helper for testing checked_headers

This allows the plugin writer to create a test with only minimal setup
validating that they have created the write table in the right schema
for their plugin.

paytonrules committed:
Remove unused NewTestDB call

paytonrules committed:
Don't include any headers before starting value

This was ignored on one half of the constructed query. I fixed it by
adding a few tests and pulling most of the query into a CTE.

paytonrules committed:
Fix the contract_watcher integration test

The DB setup was broken, unintentionally.

rmulhol commented on cmd/execute.go:
```diff
@@ -96,7 +98,11 @@ func executeTransformers() {
-		extractor := logs.NewLogExtractor(&db, blockChain)
+		repo, repoErr := repositories.NewCheckedHeadersRepository(&db, genConfig.Schema)
+		if repoErr != nil {
+			LogWithCommand.Fatalf(fmt.Errorf("failed to create checked headers repository: %w", repoErr).Error())
```
nbd but if we're just logging the error and not returning it, we could probably simplify this as:
```
LogWithCommand.Fatalf("failed to create checked headers repository: %s", repoErr.Error())
```

paytonrules replied:
👍 - you have me trained to put fmt.Errorf everywhere!

rmulhol commented on cmd/backfillEvents.go:
```diff
@@ -50,7 +51,11 @@ func backFillEvents() error {
-	extractor := logs.NewLogExtractor(&db, blockChain)
+	repo, repoErr := repositories.NewCheckedHeadersRepository(&db, genConfig.Schema)
+	if repoErr != nil {
+		return fmt.Errorf("error creating checked headers repository %w", repoErr)
```
I _think_ this is probably a non-issue since we'll be running one container per schema, but maybe it'd be helpful to include the schema name in this formatted error?

paytonrules replied:
it should be in the error message, but it also shouldn't ever really happen so it won't hurt to stick in as well.

rmulhol commented on libraries/shared/test_data/test_helpers.go:
```diff
@@ -88,3 +89,30 @@ func getRandomAddress() string {
 	return stringHash[:42]
 }
+
+/// DescribeAValidCheckedHeadersModelWithSchema is provided so that
```
seems like this comment is lying about the function name 🕵️

paytonrules replied:
Well at one time it was right. It's lying about how to use it too, so I'm fixing it.

rmulhol commented on libraries/shared/test_data/test_helpers.go:
```diff
@@ -88,3 +89,30 @@ func getRandomAddress() string {
+
+	checkedHeaderRepo, repoErr := repositories.NewCheckedHeadersRepository(db, schema)
+	Expect(repoErr).NotTo(HaveOccurred())
+	uncheckedHeaders, uncheckedHeaderErr := checkedHeaderRepo.UncheckedHeaders(0, -1, 1)
```
this is totally not your fault because it's how the function is implemented - but it might be cool to wrap `-1` (and potentially all of these numbers) in vars so that the intention is more expressive

paytonrules replied:
I might - although TBH I found some of these tests to be hard to follow because variables were in the setup and had to be understood at each level.

Maybe allowing a little WET in the tests would help with that.

rmulhol commented on pkg/datastore/postgres/repositories/checked_headers_repository.go:
```diff
@@ -17,62 +17,90 @@
-	return CheckedHeadersRepository{db: db}
+func NewCheckedHeadersRepository(db *postgres.DB, schemaName string) (*CheckedHeadersRepository, error) {
+	var sq SchemaQuery
+	err := db.Get(&sq, "SELECT schema_name FROM information_schema.schemata WHERE schema_name = $1", schemaName)
```
💪

rmulhol commented on pkg/datastore/postgres/repositories/checked_headers_repository_test.go:
```diff
@@ -30,313 +32,378 @@ import (
-			recheckCheckCount   = int64(2)
-		)
+				Expect(repo.MarkHeaderChecked(headerID)).To(Succeed())
+				Expect(repo.MarkHeaderChecked(headerID)).To(Succeed())
```
nbd but might we want a stronger assert than `Succeed()`? Seems like we could verify that the val is incremented

rmulhol replied:
lol lost in the diff, I see that the assert I'm describing is below 😅

paytonrules replied:
So I ended up changing a lot of this test because I learned to features of Gomega that I don't see a lot of in our tests.  It's documented here: 

https://onsi.github.io/gomega/#handling-errors

The short version is if you return value, err from a function (as you often do from Go funcs) and you use To(Equal(blah)) that ALSO validates that any other values in the tuple are nil. So you don't have to constantly check for error.

In addition above `Expect(repo.MarkHeaderChecked(headerID)).To(Succeed())` - requires all tuples to be nil returned from `MarkHeaderChecked`. This works best for functions like `MarkHeaderChecked` that only return an error.

I really like the first bit (we could remove a lot of explicit error checking, as it's redundant). The second I like for setup/action steps like MarkHeaderChecked is here, but I don't love it, and I'm really using it more like a traditional assert where something has gone horribly wrong if it fails.

Anyway I'm highlighting this because most of our tests aren't written this way, but I'll probably start migrating tests over to use less assertions when appropriate unless there's an objection.

rmulhol replied:
👍 absolutely love it

rmulhol reviewed (approved):
Really awesome work here! None of my comments are merge blocking :shipit:

paytonrules committed:
Put schema names in the log errors

paytonrules committed:
Fix comment and doc

elizabethengelman reviewed (approved):
👏
